Support proxying client requests through direct and tunneling HTTP/HTTPS proxies

MisterDA opened the pull request:
This PR implements a new `Cohttp_lwt_unix.Connection_proxy` module satisfying `Cohttp_lwt.S.Connection_cache`, allowing client HTTP/HTTPS requests to be routed through proxies.

I recommend a quick glance at [Everything curl - 9.8 Proxies](https://everything.curl.dev/usingcurl/proxies/index.html) for a nice introduction to proxies. As curl is well established, I try following the same terminology and support similar features.

- HTTP proxy

  > An HTTP proxy is a proxy that the client speaks HTTP with to get the transfer done.

  If the remote server is reachable through HTTP, then the client should send HTTP requests where targets are in _absolute form_ (see [RFC 9112](https://www.rfc-editor.org/rfc/rfc9112#name-absolute-form)).
  
  > When making a request to a proxy, other than a CONNECT or server-wide OPTIONS request (as detailed below), a client MUST send the target URI in "absolute-form" as the request-target.

  > `GET http://www.example.org/pub/WWW/TheProject.html HTTP/1.1`
  
  The first step is to thread an `~absolute_form` parameter when building requests. This way, the `Cohttp_lwt_unix.Connection_cache` module can simply be extended with an optional `?proxy` parameter, taking the URI of the proxy. If a proxy is specified, all requests will be send in the absolute form, and instead of connecting to the remote server, the client will transparently connect to the proxy. Conduit handles the connection to an HTTPS proxy automatically.
  
- [HTTPS with HTTP proxy](https://everything.curl.dev/usingcurl/proxies/http.html#https-with-http-proxy)

  If the remote server uses HTTPS, then a tunneling proxy needs to be used. The CONNECT HTTP requests asks the proxy to open a connection to the remote server. If it succeeds, the proxy then blindly forwards data between the client and the remote server. Conduit needs to be modified to handle an HTTPS connection (more specifically, a TLS handshake and encryption layer) on top of an already opened connection. If the proxy itself is reachable through HTTPS, conduit needs to be able to tunnel a TLS connection on top of another TLS connection.
  
  To handle tunneled client connections, we provide the (private) `Cohttp_lwt.Connection_tunnel`, which takes the proxy URI when created. It then maintains a cache of opened connections via the proxy to each remote server, and is able to distribute new requests amongst currently opened connections (the code is mostly copied from `Cohttp_lwt.Connection_cache` balancing). 
  
  We expose the `Cohttp_lwt.Connection_proxy` via `Cohttp_lwt_unix.Connection_proxy` which, based on the remote URI, selects either a direct (absolute-form queries) or tunneled connection.
  
- [MITM proxy](https://everything.curl.dev/usingcurl/proxies/mitm.html)
  
  > [MITM proxies] require users to install a custom "trust root" (Certificate Authority (CA) certificate) in the client, and then the proxy terminates all TLS traffic from the client, impersonates the remote server and acts like a proxy.
  
  This can be achieved by installing a new certificate system-wide, or passing it to the ca-certs library via external means (environment variable, parameter at startup, ...). See https://github.com/mirage/ca-certs/pull/30.
  
- [Proxy environment variables](https://everything.curl.dev/usingcurl/proxies/env.html)

  We show how to support `ALL_PROXY`, `NO_PROXY`, and `<scheme>_proxy` (such as `http_proxy`, `https_proxy`), as documented by the curl project.
  
This PR relies on tunneling the TLS-on-TLS feature being added to Conduit. We've added some opam pin-depends to allow it to build with the latest dependencies.

MisterDA committed:
http: support absolute-form URI for requests to proxies

> When making a request to a proxy, other than a CONNECT or
> server-wide OPTIONS request (as detailed below), a client MUST send
> the target URI in "absolute-form" as the request-target.

https://www.rfc-editor.org/rfc/rfc9112#name-absolute-form

See https://www.rfc-editor.org/rfc/rfc3986#appendix-A for the ABNF of
absolute-URI.

Signed-off-by: Antonin Décimo <antonin@tarides.com>

MisterDA committed:
cohttp-lwt: dispatch requests through direct and tunneling proxies

To ensure end-to-end security it's possible to use a tunneling
proxy. The proxy in the middle blindly forwards data.

This is needed if the remote server is available via HTTPS. First, a
CONNECT request is made to the proxy with the remote server as
target. If it succeeds, a new connection can be made to the remote
server, tunneled via the connection made to the proxy.

See also RFC 9110 § 9.3.6. CONNECT.
https://www.rfc-editor.org/rfc/rfc9110#name-connect

We consider that it's a sane default to always tunnel connections to
HTTPS remote server. We provide the Connection_proxy module that
automatically opens connections to a direct proxy or a tunneling
proxy, based on the remote sheme used.

We show how to respect curl's [scheme]_proxy, ALL_PROXY, and NO_PROXY
environment variables.
https://curl.se/libcurl/c/libcurl-env.html

Signed-off-by: Antonin Décimo <antonin@tarides.com>

MisterDA committed:
cohttp-lwt: allow connections to use absolute-form URI for proxies

Signed-off-by: Antonin Décimo <antonin@tarides.com>

MisterDA committed:
cohttp-lwt: support client using direct proxies

Signed-off-by: Antonin Décimo <antonin@tarides.com>

MisterDA committed:
cohttp-lwt: add a Connection_cache module interface

Signed-off-by: Antonin Décimo <antonin@tarides.com>

MisterDA committed:
cohttp-lwt-unix: example client supporting proxies

Signed-off-by: Antonin Décimo <antonin@tarides.com>

art-w commented on cohttp-lwt/src/connection_cache.ml:
```diff
@@ -168,3 +153,323 @@ end = struct
+            | Name _ -> false
+            | Ipaddr_prefix network ->
+                let subnet = Ipaddr.Prefix.of_addr hostip in
+                Ipaddr.Prefix.subset ~subnet ~network)
```
`Ipaddr.Prefix.mem`?

art-w commented on cohttp-lwt-unix.opam:
```diff
@@ -59,3 +59,10 @@ build: [
     "@doc" {with-doc}
   ]
 ]
+pin-depends: [
+  [ "tls.0.17.5" "git+https://github.com/MisterDA/ocaml-tls.git#tunnel" ]
+  [ "conduit.6.2.3" "git+https://github.com/art-w/ocaml-conduit.git#tunnel" ]
+  [ "conduit-lwt.6.2.3" "git+https://github.com/art-w/ocaml-conduit.git#tunnel" ]
+  [ "conduit-lwt-unix.6.2.3" "git+https://github.com/art-w/ocaml-conduit.git#tunnel" ]
+  # [ "ca-certs.0.2.3" "git+https://github.com/MisterDA/ca-certs.git#extra-ca" ]
+]
```
```suggestion
pin-depends: [
  [ "tls.dev"                    "git+https://github.com/art-w/ocaml-tls.git#no-cstruct_tunnel" ]
  [ "tls-lwt.dev"                "git+https://github.com/art-w/ocaml-tls.git#no-cstruct_tunnel" ]
  [ "conduit.dev"                "git+https://github.com/art-w/ocaml-conduit.git#tunnel" ]
  [ "conduit-lwt.dev"            "git+https://github.com/art-w/ocaml-conduit.git#tunnel" ]
  [ "conduit-lwt-unix.dev"       "git+https://github.com/art-w/ocaml-conduit.git#tunnel" ]
  [ "ca-certs.dev"               "git+https://github.com/art-w/ca-certs.git#extra-ca" ]

  # Cstruct patches
  [ "mirage-crypto.dev"          "git+https://github.com/mirage/mirage-crypto.git#main" ]
  [ "mirage-crypto-ec.dev"       "git+https://github.com/mirage/mirage-crypto.git#main" ]
  [ "mirage-crypto-pk.dev"       "git+https://github.com/mirage/mirage-crypto.git#main" ]
  [ "mirage-crypto-rng.dev"      "git+https://github.com/mirage/mirage-crypto.git#main" ]
  [ "mirage-crypto-rng-lwt.dev"  "git+https://github.com/mirage/mirage-crypto.git#main" ]
  [ "dns.dev"                    "git+https://github.com/hannesm/ocaml-dns.git#no-cstruct" ]
  [ "dns-client.dev"             "git+https://github.com/hannesm/ocaml-dns.git#no-cstruct" ]
  [ "dns-client-lwt.dev"         "git+https://github.com/hannesm/ocaml-dns.git#no-cstruct" ]
  [ "pbkdf.dev"                  "git+https://github.com/abeaumont/ocaml-pbkdf.git#master" ]
  [ "tcpip.dev"                  "git+https://github.com/mirage/mirage-tcpip.git#randomconv-string" ]
]
```

art-w committed:
update changelog

art-w committed:
fix CI: use sexplib0 instead of sexplib

art-w committed:
fix CI: cohttp-eio, breaking change in tls.1.0.0

art-w committed:
fix CI: opam lowerbounds lint

art-w committed:
fix CI: disable testing on 4.14 since ppx_expect.v0.17.0 requires ocaml >=5.1

art-w committed:
fix CI: cohttp-async s3_cp example, breaking change in mirage-crypto

MisterDA committed:
fix CI: update nix flake.lock

rgrinberg committed:
Merge pull request #1084 from art-w/ci-fix

chore: fix CI

MisterDA committed:
Test ability to parse a proxy'ed GET request (#1089)

dependabot[bot] committed:
chore(deps): bump cachix/install-nix-action from 27 to 28 (#1088)

Bumps [cachix/install-nix-action](https://github.com/cachix/install-nix-action) from 27 to 28.
- [Release notes](https://github.com/cachix/install-nix-action/releases)
- [Commits](https://github.com/cachix/install-nix-action/compare/v27...V28)

---
updated-dependencies:
- dependency-name: cachix/install-nix-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

talex5 committed:
cohttp-eio: Use system authenticator in example

dependabot[bot] committed:
chore(deps): bump cachix/install-nix-action from V28 to 30 (#1092)

Bumps [cachix/install-nix-action](https://github.com/cachix/install-nix-action) from V28 to 30. This release includes the previously tagged commit.
- [Release notes](https://github.com/cachix/install-nix-action/releases)
- [Commits](https://github.com/cachix/install-nix-action/compare/V28...v30)

---
updated-dependencies:
- dependency-name: cachix/install-nix-action
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

talex5 committed:
Merge pull request #1091 from talex5/tls-example

cohttp-eio: Use system authenticator in example

avsm committed:
bytebuffer: no need for the buf bytes field to be mutable

avsm committed:
cohttp-mirage/lwt-jsoo: remove unused module parameter names in signatures

avsm committed:
+dune.lock in gitignore

avsm committed:
dune: bump dune lang to 3.8 to use a more modern mdx

avsm committed:
release 6.0.0

avsm committed:
github: brew update to avoid pkgconf failure

avsm committed:
ocamlformat

avsm committed:
Merge pull request #1096 from avsm/pre-6

6.0.0 release

art-w committed:
unused functor param

art-w committed:
fix opam dep

anmonteiro committed:
chore(nix): update flakes (#1100)

art-w committed:
fix opam lint

art-w committed:
opam upperbounds on dependencies
